### PR TITLE
Allow AbstractVectors as input for initial

### LINKED ISSUE #4025: initial() will not take a weight vector in QQ.
Bug found where initial() will not take a QQFieldElem; works instead with Rational{Int}.

Code demonstrating bug:
```
julia> using Oscar

julia> T = tropical_semiring()
Min tropical semiring

julia> K, t = rational_function_field(QQ, "t")
(Rational function field over QQ, t)

julia> nu = tropical_semiring_map(K, t)
Map into Min tropical semiring encoding the t-adic valuation on Rational function field over QQ

julia> R, (x1, x2, x3) = K["x1", "x2", "x3"]
(Multivariate polynomial ring in 3 variables over K, AbstractAlgebra.Generic.MPoly{AbstractAlgebra.Generic.RationalFunctionFieldElem{QQFieldElem, QQPolyRingElem}}[x1, x2, x3])

julia> initial(x1, nu, Rational{Int}[0, 0, 0])
x1

julia> initial(x1, nu, QQFieldElem[0, 0, 0])
ERROR: Cannot promote to common type
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] promote(x::TropicalSemiringElem{typeof(min)}, y::QQFieldElem)
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/tABFQ/src/NCRings.jl:54
 [3] *(x::TropicalSemiringElem{typeof(min)}, y::QQFieldElem)
   @ AbstractAlgebra ~/.julia/packages/AbstractAlgebra/tABFQ/src/NCRings.jl:80
 [4] (::Oscar.var"#5755#5757"{…})(::Tuple{…})
   @ Oscar ./none:0
 [5] iterate
   @ ./generator.jl:47 [inlined]
 [6] collect
   @ ./array.jl:834 [inlined]
 [7] initial(f::AbstractAlgebra.Generic.MPoly{…}, nu::TropicalSemiringMap{…}, w::Vector{…}; perturbation::Nothing)
   @ Oscar ~/.julia/dev/Oscar/src/TropicalGeometry/initial.jl:57
 [8] initial(f::AbstractAlgebra.Generic.MPoly{…}, nu::TropicalSemiringMap{…}, w::Vector{…})
   @ Oscar ~/.julia/dev/Oscar/src/TropicalGeometry/initial.jl:49
 [9] top-level scope
   @ REPL[10]:1
Some type information was truncated. Use `show(err)` to see complete types.

```

On-System Oscar version information:

```
julia> Oscar.versioninfo(full=true)
OSCAR version 1.2.0-DEV - #master, 2008cc6413 -- 2024-08-13 20:56:24 +0200
  combining:
    AbstractAlgebra.jl   v0.42.0
    GAP.jl               v0.10.4
    Hecke.jl             v0.33.2
    Nemo.jl              v0.46.0
    Polymake.jl          v0.11.20
    Singular.jl          v0.23.4
  building on:
    FLINT_jll               v300.100.300+0
    GAP_jll                 v400.1200.200+9
    Singular_jll            v404.0.301+0
    libpolymake_julia_jll   v0.12.1+0
    libsingular_julia_jll   v0.45.2+0
    polymake_jll            v400.1200.1+0
See `]st -m` for a full list of dependencies.

Julia Version 1.10.4
Commit 48d4fd48430 (2024-06-04 10:41 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, apple-m1)
Threads: 1 default, 0 interactive, 1 GC (on 4 virtual cores)
Official https://julialang.org/ release

```

### PR DESCRIPTION
**Pull request by YueRen**
Fixes https://github.com/oscar-system/Oscar.jl/issues/4025.

Important because functions such as `relative_interior_point` return PointVectors, and points on a tropical variety or in a Groebner complex are primary candidates for weight vectors.

### CODE REVIEW

**Review by benlorenz — DISMISSED** [review]
I think most of the changes are unneccessary if you would change the relevant signatures (for `groebner_basis` and `initial`) to allow `w::AbstractVector{...}` instead of `w::Vector{...}`.

What is still needed is the `tropical_semiring(nu)(...)`, otherwise this leads to some promotion error, but that also happens with `Vector{QQFieldElem}`.

**Comment by YueRen** [comment]
> I think most of the changes are unneccessary if you would change the relevant signatures (for `groebner_basis` and `initial`) to allow `w::AbstractVector{...}` instead of `w::Vector{...}`.
> 
> What is still needed is the `tropical_semiring(nu)(...)`, otherwise this leads to some promotion error, but that also happens with `Vector{QQFieldElem}`.

Good point, I've updated the pull request.

**Comment by benlorenz** [comment]
Somewhat related question that I just noticed while looking at the code here:
https://github.com/oscar-system/Oscar.jl/blob/988b6077c9f1c272f06d5c113e98dfca7b2cfa5d/src/TropicalGeometry/initial.jl#L64-L70

I don't see the contents of `perturbation` being used at all (except for that `isnothing` check) ?

**Comment by YueRen** [comment]
> Somewhat related question that I just noticed while looking at the code here:
> 
> https://github.com/oscar-system/Oscar.jl/blob/988b6077c9f1c272f06d5c113e98dfca7b2cfa5d/src/TropicalGeometry/initial.jl#L64-L70
> 
> I don't see the contents of `perturbation` being used at all (except for that `isnothing` check) ?

Good catch, I think that part is obsolete and can be safely removed.  It was originally intended for the tropical traversal algorithm, but 
(a) never used 
(b) that part can be handled cleaner simply by calling `initial` twice, imo.

**Review by benlorenz — APPROVED** [review]

File: src/TropicalGeometry/initial.jl [hunk 1]
```diff
@@ -110,7 +100,7 @@ Ideal generated by
   -2*x^2*y + 3*y^3
 ```
 """
-function initial(I::MPolyIdeal, nu::TropicalSemiringMap, w::Vector{<:Union{QQFieldElem,ZZRingElem,Rational,Integer}}; skip_groebner_basis_computation::Bool=false)
+function initial(I::MPolyIdeal, nu::TropicalSemiringMap, w::AbstractVector{<:Union{QQFieldElem,ZZRingElem,Rational,Integer}}; skip_groebner_basis_computation::Bool=false)
     G = (skip_groebner_basis_computation ? gens(G) : groebner_basis(I,nu,w))
     return ideal(initial.(G,Ref(nu),Ref(w)))
```
Comment (benlorenz) [inline src/TropicalGeometry/initial.jl]:
```suggestion
function initial(I::MPolyIdeal, nu::TropicalSemiringMap, w::AbstractVector{<:Union{QQFieldElem,ZZRingElem,Rational,Integer}}; skip_groebner_basis_computation::Bool=false)
    G = (skip_groebner_basis_computation ? gens(I) : groebner_basis(I,nu,w))
    return ideal(initial.(G,Ref(nu),Ref(w)))
```

**Comment by YueRen** [comment]
I think the pull-request is ready to be merged, unless @benlorenz has any further comments.

**Pull request merged**